### ISSUE TITLE
Use graceful shutdown as a default policy (for SIGTERM)

### ISSUE THREAD

**Issue by kwart** [hazelcast/hazelcast#11624]
Current Hazelcast Member shutdown policy uses `TERMINATE` as a default. We should consider to change it to `GRACEFUL` and allow users cleaner handling of `SIGTERM` signal.

**Background**
Hazelcast knows how to do the graceful shutdown. Nevertheless the graceful shutdown is not used as a default policy. If a user starts/stops members with our provided shell scripts, then with the default configuration the `stop.sh` terminates the Member instead of doing the graceful shutdown.

It's important to change the default behavior at least for the scripts, because for instance the [rolling upgrades requires the graceful shutdown](https://github.com/hazelcast/hazelcast-reference-manual/blame/v3.9-EA/src/RollingMemberUpgrades.md#L39) to be used.

**Suggested improvements**
* Either change the default shutdown policy from `TERMINATE` to `GRACEFUL`
  * decrease the  default for timeout if necessary (current value is `600` second)
* Or set explicitly the `GRACEFUL` policy in the scripts (`bin/start.*`, `demo/console.*`, docker start scripts)

**Comment by Donnerbart** [comment]
So `SIGKILL` (aka `-9`) for `TERMINATE` and `SIGTERM` (aka `-15`) for `GRACEFUL`? It makes sense to me. Most devops people will probably try `kill $pid` first and if that fails will automatically escalate to `kill -9 $pid`. But it would be nice to get some feedback from TechOps about this.
```
       Signal     Value     Action   Comment
       ──────────────────────────────────────────────────────────────────────
       SIGHUP        1       Term    Hangup detected on controlling terminal
                                     or death of controlling process
       SIGINT        2       Term    Interrupt from keyboard
       SIGQUIT       3       Core    Quit from keyboard
       SIGILL        4       Core    Illegal Instruction
       SIGABRT       6       Core    Abort signal from abort(3)
       SIGFPE        8       Core    Floating point exception
       SIGKILL       9       Term    Kill signal
       SIGSEGV      11       Core    Invalid memory reference
       SIGPIPE      13       Term    Broken pipe: write to pipe with no
                                     readers
       SIGALRM      14       Term    Timer signal from alarm(2)
       SIGTERM      15       Term    Termination signal
       SIGUSR1   30,10,16    Term    User-defined signal 1
       SIGUSR2   31,12,17    Term    User-defined signal 2
       SIGCHLD   20,17,18    Ign     Child stopped or terminated
       SIGCONT   19,18,25    Cont    Continue if stopped
       SIGSTOP   17,19,23    Stop    Stop process
       SIGTSTP   18,20,24    Stop    Stop typed at terminal
       SIGTTIN   21,21,26    Stop    Terminal input for background process
       SIGTTOU   22,22,27    Stop    Terminal output for background process
```

**Comment by mdogan** [comment]
Afaik a process can't handle `SIGKILL`. OS kills the process without a notice. So, it's not a `TERMINATE` in terms of Hazelcast.

**Comment by Donnerbart** [comment]
Oh, right. Here is a nice table on which signal the JVM is stopped and on which the shutdown hook is triggered: http://journal.thobe.org/2013/02/jvms-and-kill-signals.html

**Comment by tkountis** [comment]
This has been discussed again https://github.com/hazelcast/hazelcast-docker/issues/42
It sounds logical to me, but it mostly affects integrations, so lets get their input
cc. @leszko @mesutcelik @googlielmo any takes?

**Comment by mesutcelik** [comment]
1. We do enable `GRACEFUL` by default in hazelcast helm chart.
2. Docker is documented as you see with the info how to configure shutdown hook.

I have not received any complaint so far about it being GRACEFUL not default so I have no strong opinion regarding changing it.

**Comment by leszko** [comment]
I'm all for setting graceful shutdown by default. It's a good idea :)

**Comment by tkountis** [comment]
On a second/third thought on that, I am not sure whether this is the right choice. Trying to be the admin advocate here, but how often does an admin send a SIGTERM and waits for mins (maybe even hour(s), depending on cluster size) for the signal to complete? This can easily freeze terminals, or even OS shutdown process (for a given grace period). 
On the other hand, we never really got a complain about default behaviour and it is configurable with the property, as Mesut said, they already default to GRACEFUL for helm chart.
Therefore, I am a bit skeptical whether this will be an happy surprise to Hazelcast users or an unexpected one.

**Comment by tkountis** [comment]
Furthermore, the shutdown hooks run in parallel, so I am wondering whether it will be an issue for 1 of them to take so much longer to stop, while the rest (if any, registered by the app) finish faster, maybe cleaning up other resources potentially shared with Hazelcast, eg. files. 

Quoting the relevant [javadoc](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Runtime.html#addShutdownHook(java.lang.Thread))

> Shutdown hooks run at a delicate time in the life cycle of a virtual machine and should therefore be coded defensively. They should, in particular, be written to be thread-safe and to avoid deadlocks insofar as possible. They should also not rely blindly upon services that may have registered their own shutdown hooks and therefore may themselves in the process of shutting down. Attempts to use other thread-based services such as the AWT event-dispatch thread, for example, may lead to deadlocks.
> 
> Shutdown hooks should also finish their work quickly. When a program invokes exit the expectation is that the virtual machine will promptly shut down and exit. When the virtual machine is terminated due to user logoff or system shutdown the underlying operating system may only allow a fixed amount of time in which to shut down and exit. It is therefore inadvisable to attempt any user interaction or to perform a long-running computation in a shutdown hook.

Given this, I am reluctant to make this the default behavior for our shutdown hook. The correct approach would be for the app to handle its lifecycle based on its deployment/un-deployment plan and making sure the cluster is safe before calling shutdown. Having this configurable, allows already for other integrations to have separate logic based on their needs, so I am happy with that side of things too. 

My only remaining concern is the start/stop scripts. The current behavior is to rely on SIGTERM which as we said, by default terminates the instance, but the real problem is that its not verifying cluster state (ie. `isClusterSafe`). I think the right change here would be to improve the shutdown API, by rejecting shutdown requests on a cluster who is undergoing migrations unless we force it too.
@metanet @mdogan any thoughts on that?

**Comment by mdogan** [comment]
A few command line apps I had used before was doing a graceful stop at the first signal/interrupt (`ctrl + c`) but exiting immediately after second `ctrl + c` stroke. Can't we do similar?

**Comment by tkountis** [comment]
Investing some more time into this, helped me re-enforce my original opinion. 
I am personally, against any change related to changing the signal handlers we currently have, trigger a graceful shutdown. All signal handling is meant to be almost instantaneous, i.e., closing FDs and verifying a healthy process exit. A node shutdown should be triggered by the application itself, and be part of its lifecycle and not rely on OS initiated and managed signals. 

A few arguments:

- Managing SIGNALs other than the default shutdown hooks in Java, is kinda problematic. Either you go with the default shutdown hook, which is supposed to be portable on multiple platforms, or you mess around with the private Signal API which is in `sun.misc.*`  and you have to control corner cases manually, not funny! What happens when you run in a command line, and you close the window by clicking on the X button? Do we hang the process until the shutdown is finished? What happens when you shutdown the computer, is this a GRACEFUL shutdown for a distributed app?

- Running Hazelcast in embedded mode as a local cache layer (library), it doesn't really matter whether you shutdown or terminate. It shouldn't even have a shutdown hook enabled whatsoever, although we do have it enable by default. Its application logic if they want to handle shutdown or not in any form. Hazelcast in that case is a pure library.

- Running Hazelcast as a standalone process through your DEV environment, also shouldn't matter.

- Running Hazelcast as a standalone process (forming a cluster), does matter. We need to have ways to allow the user to shutdown the node. For that we have correct and incorrect approaches, always from my personal point-of-view of course. The correct ones, JMX controls and Management Center support, and on the other side, we have shell scripts, both the ones included in the zipped artifact, and the ones we provide through the hazelcast-member service, which rely on SIGTERM.

This last point, is not ideal, I totally agree on this being an issue. We should look into fixing that by having a different stop mechanism from scripts. eg. IPC to request shutdown and then trigger SIGTERM to exit? This however, is out of the scope of a GH issue, and should be turned into a PRD, together with improving the whole hazelcast-member project which feels like it needs some love.

Having said all that, I provided a PR which adds a couple of extra comments on the shutdown policy property, and introduces an extra "enhancement" for the GRACEFUL mode, to support subsequent signal handling that will interrupt the shutdown, and revert back to terminate mode. Even that change, introduces another dependency to `sun.misc.*` which I am also not happy :-) 
https://github.com/hazelcast/hazelcast/pull/16368

**TL;DR**
I think our main issue here is the bad state of things when it comes to Hazelcast IMDG and Linux native service support. We lack a healthy service integration, that controls the application. I don't think a signal or an interrupt should be triggering a graceful shutdown. I will suggest that we close this issue with or without the provided PR, and open a PRD and approach this the right way.

**Comment by mmedenjak** [comment]
Given that this has evolved into a complex topic, that the behaviour is already configurable and that we haven't received many complaints from the customers or community, moving this out of 4.0 scope into backlog.
